### ISSUE TITLE
Add Notification Center to avoid the pop up dialogs. 

### ISSUE THREAD

**Issue by juanis2112** [spyder-ide/ux-improvements#8]
We want to add a Notification Center to show all the warnings or notifications and avoiding pop up dialogs that may result disturbing for users. For this, we need also a bell to display it and "toaster menu" that shows notifications the first time.
@isabela-pf can you help us think a bit more about the design about this Notification Center?

**Comment by isabela-pf** [comment]
This is definitely a work in progress, so I don’t have a whole notification center thought out or mocked up yet. At the moment, I have ideas for toast notifications and potential toolbar reorganization that might help provide a place for a notification center. I’m going to post this across a few comments so that it is easier to navigate when I am talking about each aspect.

One of the biggest things I’m looking for is better understanding of what notifications are needed and how they are currently existing in Spyder. This will help me make smarter choices and be able to map out user flows or needs for more specific interactions.

In meeting with some of the Spyder team, some notifications might include
- Telling users there is a newer version of Spyder available
- Any errors that do not make sense in the console
- Kite permissions (or updates if relevant?)

I’d love to hear what other notifications or use cases people might have in mind.

**Comment by isabela-pf** [comment]
Right now, I’m experimenting with toast colors to fit in with the default dark mode. I’ve provided options in the same colors as my work with tooltips in #2 and am open to opinions on what color is best suited. 

These should be used to support short messages (1–3 lines of text) and a button to dismiss the toast, bring up the notification center with that notifications selected, or initiate some kind of action if relevant. I haven’t fully worked out what buttons make sense because I want a better idea of what kind of notifications might be common in order to make a smarter choice.
![Toast 2 Line 1 Button](https://user-images.githubusercontent.com/50221806/87354550-91d61b80-c513-11ea-9b90-a6111d8afe8a.png)

![Toast 3 Line 2 Button](https://user-images.githubusercontent.com/50221806/87354553-926eb200-c513-11ea-805e-1a653f18cd81.png)

![EX Light](https://user-images.githubusercontent.com/50221806/87354609-afa38080-c513-11ea-90d1-5533ad8fcbb4.png)

![EX Dark](https://user-images.githubusercontent.com/50221806/87354627-b500cb00-c513-11ea-9f0f-ae3d67ca7c50.png)

![EX Blue](https://user-images.githubusercontent.com/50221806/87354642-b7fbbb80-c513-11ea-89f8-e7bca8141a57.png)

Some other notes about toast notifications:
- They shouldn’t be permanent (or require manual dismissing). A toast is the initial notification, and a toast will always appear along with the more informative and permanent message in the notification center. I’m thinking toasts only stick around for 15 seconds without any interaction before disappearing. They can be dismissed manually earlier, though.
- They shouldn’t replace all dialogs. I think there’s more thought to be done around what kind of notifications need to be toasts (and in the notification center), but I wanted to note that just in case.

**Comment by isabela-pf** [comment]
Because notifications can be related to all levels/all panes of Spyder, it would make sense to put a notification center like this in the highest level toolbar. Since there is already limited room there, I looked closer what is in that toolbar and found that it looks like a mix of things relating to the editor pane and things that relate to Spyder overall or across multiple panes.

I think it would make more sense to treat the editor pane like all the other panes and give it it’s own toolbar while keeping the top toolbar for non-pane-specific interactions. Just in terms of organization, the split would look kind of like this:

![Reorganized Toolbar](https://user-images.githubusercontent.com/50221806/87354807-0d37cd00-c514-11ea-8bc8-3201d69dd0e5.png)

If that’s not feasible, we’ll probably need to decrease icon size in that top toolbar to fit another icon or come up with another solution.

**Comment by isabela-pf** [comment]
I’ve been thinking about what kind of content fits for this type of notification system to make sure that I have a better grasp of potential use cases and can make sure this system has everything it needs to function. Here’s what I’ve thought might be typical toast notifications so far:
- Prompts to update Spyder
- Enable third party elements (like Kite)
- First time opening prompts to set up things
- Any non-breaking whole application errors (like things that wouldn’t belong to a specific pane)
- Notification that something that was running is done

In summary, toasts should have non-urgent content with choices that do not require immediate input. They also should not replace all dialogs.

It’d be helpful to get if there’s any other types of content that should be a toast or any problems with these current ideas.

**Comment by bcolsen** [comment]
> Notification that something that was running is done

I like this. It would be great to include console name and whether the processes ended in an error or normally in this message

**Comment by isabela-pf** [comment]
I’ve mocked up some toast notifications with example content based on the last comments. There’s a few things that would be useful to review here.
- Are these the types of content you expected to see in a notification like this?
- Are these the kind of button options you think will work?
- Are they informative enough in the limited number of lines?

![Two toast notifications with the content 
"Welcome to Spyder! Explore our guided resources or get started on your own" and "A new version of Spyder is available! Update to v5.0.0 now for the latest features"](https://user-images.githubusercontent.com/50221806/88942104-a9651200-d23e-11ea-86e9-c0ebcd221c83.png)
![Two toast notifications with the content "Kite is a third party integration that improves Spyder’s autocompletion. Allow Kite?" and "A cell has been running for more than 10 minutes. Spyder may run slower as a result"](https://user-images.githubusercontent.com/50221806/88942109-ab2ed580-d23e-11ea-9c16-eac5e1d96641.png)
![Two toast notifications with the content "Is Spyder not behaving as expected? Review our resources to find the problem" and "Cell 22–35 has finished running successfully. Check Console 1/A for more details"](https://user-images.githubusercontent.com/50221806/88942112-ac600280-d23e-11ea-84bb-325530f428dc.png)

Here’s some concerns about button options I have:
- I don’t have any decline buttons because toast notifications only last 15 seconds. But then are users forced to wait 15 seconds if the notification is in their way (since it will briefly cover the upper right panes)? Or would users have to open and close the notification center to get it to go away? Both are bad experiences, but it also seems like a waste to always have one of the options just be “dismiss.”
- Alternatively, should the second button always be to take users to the notification center? I’m not sure if it’s obnoxious or helpful.

**Comment by isabela-pf** [comment]
There are a lot of questions to consider to start reviewing the notification center itself, so I’m going to be posting them across few comments to hopefully make it more manageable to read and reply.

First, here are options for how the notification center itself will look. Choices being made here are about how notifications appear (listed, still looking like toast notifications), how read/unread is indicated (dots, corners, contrast, etc.), how a notification is selected (checkbox, whole area), and where the button prompts originally seen in the toast notifications appear.

The first three options experiment mainly with ways to indicate read/unread status for notifications. They also have buttons hidden until a notification is expanded.

<img width="550" alt="28" src="https://user-images.githubusercontent.com/50221806/88976626-36c15a00-d271-11ea-9d87-f0dd939346b3.png">

<img width="550" alt="29" src="https://user-images.githubusercontent.com/50221806/88976633-37f28700-d271-11ea-81ec-4664bba9cd2c.png">

<img width="550" alt="30" src="https://user-images.githubusercontent.com/50221806/88976635-388b1d80-d271-11ea-89b9-6fd4d0c78349.png">

The below option has the same button styling as the last three, but they are shown at all times instead of on expanded notifications only.

<img width="550" alt="31" src="https://user-images.githubusercontent.com/50221806/88977335-92401780-d272-11ea-81f4-75ad3147e242.png">

This final option preserves the look of toast notifications, just groups them together in this interface. I haven't worked  an expanded/more info form in this option yet, but if there is interest in exploring this direction more I am happy to. Depending on notifications, this might not be needed on any of the options.

<img width="550" alt="32" src="https://user-images.githubusercontent.com/50221806/88977684-3629c300-d273-11ea-8ab0-94a7f734faef.png">

**Comment by isabela-pf** [comment]
There are two places I think the notification center could reasonably appear: as an overlay from the main toolbar (like the quick switch between files) or as another pane in the upper right region. I prefer not to make another pane—especially since notifications can relate to any pane or even be for the application overall—but wanted to provide another option in case there was a reason that overlay treatment is not feasible.

As a main toolbar linked overlay:

<img width="1344" alt="overlay" src="https://user-images.githubusercontent.com/50221806/88978313-5017d580-d274-11ea-87f2-e1966e9b0982.png">

I find it hard to see in this mockup even though it is designed closely to the existing quick switch between file interface. I would definitely want to work out ways to help it stand out more if people like the concept of this direction.

As a pane in the upper right:

<img width="829" alt="pane" src="https://user-images.githubusercontent.com/50221806/88978318-54dc8980-d274-11ea-97ed-0daf834d2dda.png">

**Comment by isabela-pf** [comment]
I wanted to note a few other options to consider. The actions I think are absolutely necessary in the toolbar are mark as read, mark as unread, delete/clear, and select all. Here, I mocked up an option below that includes a sort by recent (default) or sort by unread first because I could imagine that as a useful feature that might not take too much work to implement.

<img width="550" alt="27" src="https://user-images.githubusercontent.com/50221806/88978388-79d0fc80-d274-11ea-8d4f-30c81c0efae4.png">

I have not worked out where the notifications icon button should go in the main toolbar since I have not played more with breaking up the large number of buttons currently there. I am planning to use the [Material icon “notifications.”](https://material.io/resources/icons/?search=notif&icon=notifications&style=baseline)

**Comment by juanis2112** [comment]
We like option #4. Some general comments from our meeting:

- [ ] Add a bell icon (with red dot NO NUMBERS) for displaying notifications (On the right of the arrow that points up).
- [ ] Narrow stripe from top to bottom (similar to Mac notifications)
- [ ] Add Scrollbar 
- [ ] Behavior: Popping up similar to the file explorer
- [ ] Add toast notifications for the first time they pop. 
- [ ] No read/unread state
- [ ] No checkboxes 
- [ ] No “Select all”
- [ ] Only thing in toolbar will be the trash can for deleting all the notifications at once
- [ ] Add an x in the top corner of each notification to delete it.
- [ ] The date should go below the notification (not at the right)
- [ ] The buttons that call to action should be in the side displayed by default (in case of needed)  
- [ ] Wait for color palette

**Comment by juanis2112** [comment]
For the toast notifications, we like the position of the 1st screenshots (top right) but allow to customise position.

**Comment by juanis2112** [comment]
Take into account notification preferences in the Preferences dialog
